File: packages/server/src/modules/notifications/models/notification.ts
import { ObjectId } from 'mongodb';
import { Actor } from '@/modules/actors/models/actor';
import { Post } from '@/modules/posts/models/post';
import { Comment } from '@/modules/comments/models/comment';

/**
 * Notification types supported by the system
 */
export enum NotificationType {
  FOLLOW = 'follow',
  LIKE = 'like',
  COMMENT = 'comment',
  MENTION = 'mention',
  REPOST = 'repost',
}

/**
 * Notification interface
 */
export interface Notification {
  _id?: string | ObjectId;
  recipientUserId: string;
  actorUserId?: string;
  type: NotificationType;
  postId?: string;
  commentId?: string;
  read: boolean;
  createdAt: Date;
  updatedAt: Date;
}

/**
 * Data needed to create a notification
 */
export interface CreateNotificationDto {
  recipientUserId: string;
  actorUserId?: string;
  type: NotificationType;
  postId?: string;
  commentId?: string;
  read?: boolean;
  content?: string;
}

/**
 * Formatted notification with actor and content details
 */
export interface FormattedNotification extends Notification {
  id: string;
  type: NotificationType;
  actor?: {
    id: string;
    username: string;
    displayName?: string;
    avatarUrl?: string;
  };
  postId?: string;
  commentId?: string;
  content?: string;
  read: boolean;
  createdAt: Date;
}

File: packages/ios/Saturn/Models/Notification.swift
import Foundation

enum NotificationType: String, Codable, CaseIterable {
    case LIKE
    case FOLLOW
    case MENTION
    case REPOST
    case COMMENT
}

struct NotificationPost: Codable, Identifiable {
    let id: String
    let content: String
}

struct PaginationInfo: Codable {
    let currentPage: Int
    let totalPages: Int
    let totalItems: Int
    let hasNext: Bool
    let hasPrev: Bool
    
    enum CodingKeys: String, CodingKey {
        case currentPage = "current_page"
        case totalPages = "total_pages"
        case totalItems = "total_items"
        case hasNext = "has_next"
        case hasPrev = "has_prev"
    }
}

struct Notification: Codable, Identifiable {
    let id: String
    let type: NotificationType
    let read: Bool
    let createdAt: String
    let actor: Actor
    let post: NotificationPost?
    
    enum CodingKeys: String, CodingKey {
        case id
        case type
        case read
        case createdAt = "created_at"
        case actor
        case post
    }
}

struct NotificationResponse: Codable {
    let notifications: [Notification]
    let pagination: PaginationInfo
}